### ISSUE TITLE
Optional arguments don't work in exported functions called from Excel worksheet

### ISSUE THREAD

**Issue by leo-liar** [sdllc/Basic-Excel-R-Toolkit#113]
System:
BERT v2.3.4
Windows 8.1 Pro x64
Excel 2016 (MSO 365) v16.0.9126.2116

---

While playing around and trying some things with the provided example functions I noticed an error in the handling of optional function parameters. For easy reproduction I wrote this little test function:

```R
# 
# simple logical AND for testing
#
TestAND <- function( arg0 = TRUE, arg1 = TRUE, arg2 = TRUE, arg3 = TRUE ) {
  arg0 && arg1 && arg2 && arg3
}
```

When using this in Excel there's an error if the second or (in general) one of the middle optional arguments is left out, as shown in the attached [**example file**][file].

![TestAND.xlsx][pic]

I first became aware of this while using the [**Matrix Correlation**][1] [**Example Spreadsheet**][2], found in the [documentation][3] of BERT, where this error occurs in the resulting correlation matrix for _Data in columns_.

In both cases the R console complains about _invalid argument types_.
The shell output for **TestAND** when omitting middle arguments reads `Error in arg0 && arg1 : invalid 'y' type in 'x && y'` while the shell output for **MatrixCorrelation** (which for the first correlation matrix omits parameters by default) is `Error in !data.in.columns : invalid argument type`.

Quick testing a self-written sum function with simple integer default values gave no errors but in the [**Box-Cox Transform**][4] [**example**][5] the same erroneous behaviour can be observed if `range.min` and/or `range.max` are not defined by the function call from Excel. The shell output here reads as follows:

```R
Error in seq.default(range.min, range.max, by = range.step) :
  'from' muss Länge 1 haben
```

To easily reproduce this bug I have attached screenshots depicting the error for all mentioned cases below.

<details><summary>Attachments</summary>

- [Matrix Correlation][matrixcorrel]
![matrixcorrelation-example.xlsx][matrixcorrel]
- [Box-Cox Transform][boxcox]
![boxcox-example.xlsx][boxcox]
- [Shell output][shelloutput]
![BERT console R shell][shelloutput]
- [TestAND.xlsx][file]
</details>

---

<details><summary>Translations</summary>

German | English
:---: | :---:
`WAHR` | `TRUE`
`FALSCH` | `FALSE`
`#WERT!` | `#VALUE!`
`'from' muss Länge 1 haben` | `'from' must be of length 1`
</details>

[1]: https://bert-toolkit.com/examples/matrixcorrelation.R "matrixcorrelation.R"

[2]: https://bert-toolkit.com/examples/matrixcorrelation-example.xlsx "matrixcorrelation-example.xlsx"

[3]: https://bert-toolkit.com/bert-example-functions "BERT Example Functions"

[4]: https://bert-toolkit.com/examples/boxcox.R "boxcox.R"

[5]: https://bert-toolkit.com/examples/boxcox-example.xlsx "boxcox-example.xlsx"

[file]: https://github.com/sdllc/Basic-Excel-R-Toolkit/files/1913504/TestAND.xlsx "TestAND.xlsx"

[pic]: https://user-images.githubusercontent.com/25538015/38776361-814b7a4e-4095-11e8-951b-9fd8dbfe9cd1.png "TestAND.xlsx"

[matrixcorrel]: https://user-images.githubusercontent.com/25538015/38777164-becf9ed2-40a3-11e8-8caa-77b287de22af.png "matrixcorrelation-example.xlsx"

[boxcox]: https://user-images.githubusercontent.com/25538015/38777179-0767393e-40a4-11e8-87ca-69d1639084e0.png "boxcox-example.xlsx"

[shelloutput]: https://user-images.githubusercontent.com/25538015/38777180-0f48e422-40a4-11e8-9670-3a3992fc0eb8.png "BERT console R shell"

**Comment by duncanwerner** [comment]
Sounds like we're not handling the Missing type properly. I'll have a look.

**Comment by duncanwerner** [comment]
So the reason this is happening is that we construct R calls with `do.call`, which takes a list of arguments. There is no way, as far as I can tell, to insert an "empty" or "missing" value. What we were doing was setting `R_NilValue`, which translates to `NULL` on the R side.

I think there are a few different ways that we could resolve this, but it seems like the one with the least overhead is to use `do.call` and name the arguments in the list. That way we can drop missing arguments  from the list, without relying on position.

This will still support `...` arguments, but only if the `...` is the last argument in the list (although a three-dots argument that is not the last argument would not work properly in Excel anyway).

**Comment by duncanwerner** [comment]
Another option would be to explicitly set default values (we know what they are), but that would only work for scalars or things we could explicitly translate.

**Comment by leo-liar** [comment]
Well, I'm by no means an expert but the first solution you proposed sounds reasonable. Especially if the only trade-off is just `...` arguments having to be placed after all other arguments which Excel requires either way.

**Comment by leo-liar** [comment]
Ah, interesting. Thanks again for the comprehensive answer. The advantage of having the info about a default value visible is (in my opinion) that it's immediately discernible that said argument/parameter is optional.

Am I understanding the [`linked commit`][commit2] and your explanation right that default values will now be presented regardless of their type? If so, then I can't await to see this included in a release!

Either way, since my original problem has already been solved with [`9fc89e1`][commit1] in [release 2.3.6][release] I'm going the close this issue shortly. That is unless there's something important left you want to add here. And again, thank you a lot.

[commit2]: sdllc@ba4fd0fe02807136f7e1b700ea33d5b3d455c024 "fix default argument handling #113"
[commit1]: sdllc@9fc89e1af7b96cd1241fe41012fe63d67f2c8f5c "name arguments in R call to support intervening missing arguments, per #113"
[release]: https://github.com/sdllc/Basic-Excel-R-Toolkit/releases/tag/2.3.6 "Updates and Fixes"
